Alex/access control list

larzhole opened the pull request:
## Description
Added an Access Control List (ACL) to manage access to the API based on groups
Created new tables: "roles", and "userRoles"
Added an ACL object (src/auth/acl.ts) to configure role based access
Added ACL middleware logic and separated middleware logic under /src/middleware
Added CRUD endpoints for Roles and PATCH endpoint for UserRoles

**Issue(s):**
N/A

ericdowell commented on src/app.ts:
```diff
@@ -134,6 +137,50 @@ const registerMiddleware = (app: Express): void => {
+      // TODO possibly only read the file once
+      const aclRoles = yaml.safeLoad(fs.readFileSync('acl.yaml', 'utf-8')) as Role[]
+
+      const hasPermissions = aclRoles.filter((value: Role) => currentUserRoles.indexOf(value.name) > -1
+        && method in value
+        && value[method].includes(url)
+        ).length > 0
+
+      if (!hasPermissions) {
+        return res.sendStatus(401)
+      }
```
I would rewrite this whole thing to return early with check on `currentUserRoles` and then store `role.routes[method]` in a variable to make typescript happy.

Better to name filter argument `role` instead of `value`, easier to read within the multiple boolean `&&`'s that way. Also you can use [Array.prototype.includes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) instead of `indexOf`

We can also reverse the variable meaning to be negative instead, `isUnauthorized` and thus check for `.length === 0`.

Also we should use `status(401).json()` providing a message that's meaningful. I've started to use `'This action is unauthorized.'` for 401 responses. Also key should be `message` instead of `error`.

```suggestion
      const isUnauthorized =
        aclRoles.filter((role: Role) => {
          if (!currentUserRoles.includes(role.name)) {
            return false
          }
          const routes = role.routes[method]
          return Array.isArray(routes) && typeof routes !== 'undefined' && routes.includes(url)
        }).length === 0

      if (isUnauthorized) {
        return res.status(401).json({ message: 'This action is unauthorized.' })
      }
```

larzhole replied:
Complete

ericdowell commented on src/app.ts:
```diff
@@ -134,6 +137,50 @@ const registerMiddleware = (app: Express): void => {
+
+    } catch (error) {
+      logger.error(error)
+      return res.sendStatus(500).json({ error: 'something went wrong' })
```
You don't want to use `sendStatus`, we want to use `status`. Also I've since updated this to be `message` instead of `error` and updated the casing a little:
```suggestion
      return res.status(500).json({ message: 'Something went wrong' })
```
Again I plan on having these all in a central place so we don't need to maintain remembering silly string values and object keys.

larzhole replied:
Now that we don't need to read a file, I'm removing the try/catch so I'm deleting this line

ericdowell commented on src/app.ts:
```diff
@@ -134,6 +137,50 @@ const registerMiddleware = (app: Express): void => {
   })
+
+  // Access Control List
+  app.use(async (req: Request, res: Response, next: NextFunction) => {
```
I'm undecided if we should be apply rules globally like this or at the route level. Since we're not a big framework and things are predictable we can stay with this for now.

If something is working right we can move to creating common middleware we can apply to each route as needed.

ericdowell replied:
Based on discussion it will be better to have this placed on a route by route level. Main takeaway is we will not have logic to handle returning a 404 vs a 401.

ericdowell commented on package.json:
```diff
@@ -62,6 +62,7 @@
     "express-session": "^1.17.1",
     "express-validator": "^6.6.0",
     "helmet": "^3.23.3",
+    "js-yaml": "^3.14.0",
```
We can remove this dependencies since we can just use Javascript Objects to create our ACL Roles just as easily.
```
npm remove js-yaml @types/js-yaml
```

larzhole replied:
Complete

larzhole commented on acl.yaml:
```diff
@@ -0,0 +1,15 @@
+- name: user
```
Complete

larzhole commented on src/app.ts:
```diff
@@ -134,6 +137,50 @@ const registerMiddleware = (app: Express): void => {
+
+    try {
+      // TODO possibly only read the file once
+      const aclRoles = yaml.safeLoad(fs.readFileSync('acl.yaml', 'utf-8')) as Role[]
```
Complete

larzhole commented on src/app.ts:
```diff
@@ -134,6 +137,50 @@ const registerMiddleware = (app: Express): void => {
+
+  // Access Control List
+  app.use(async (req: Request, res: Response, next: NextFunction) => {
+    const postWhitelist = [
```
Complete

larzhole commented on src/app.ts:
```diff
@@ -134,6 +137,50 @@ const registerMiddleware = (app: Express): void => {
+    const url = req.url as Routes
+    const method = req.method.toLowerCase()
+
+    if (!user && 'post' === method && postWhitelist.includes(url)) {
```
Complete

larzhole commented on src/app.ts:
```diff
@@ -134,6 +137,50 @@ const registerMiddleware = (app: Express): void => {
+      Routes.RESET_GUEST_PASSWORD
+    ]
+    const user = req.user
+    const url = req.url as Routes
```
Complete

larzhole commented on src/types/acl-schema.d.ts:
```diff
@@ -0,0 +1,7 @@
+export interface Role {
+  name: string
+  delete?: string[]
```
Complete

larzhole commented on src/controllers/role.ts:
```diff
@@ -0,0 +1,24 @@
+import { logger } from '../logger'
+import { Role } from '../entity'
+
+export const postAddRole = async (req: Request, res: Response): Promise<Response> => {
```
@ericdowell Do you want all the CRUD + PATCH endpoints in one file here? or separate files?

ericdowell replied:
CRUD in one controller for `Role` model and PATCH for assigning user can be placed in `src/controllers/user.ts` for now.

Path can be `user/role/assign`, inputs can be placed in the JSON body:
```json
{ "userId": 1, "roleId": 1 }
```

larzhole replied:
Complete

ericdowell commented on src/migration/1594835753320-CreateUserRolesTable.ts:
```diff
@@ -0,0 +1,55 @@
+            name: 'roleId',
+            type: 'int',
+            unsigned: true,
+            isPrimary: true,
```
Is setting both of these correctly setting up a complex unique across these two columns?

If not, please look into how that can be done with TypeORM migrations. Let me know if you have questions around this.

larzhole replied:
I believe so...
I'm unable to insert duplicates into the table

ericdowell commented on src/migration/1594835746397-CreateRolesTable.ts:
```diff
@@ -0,0 +1,32 @@
+            isNullable: false,
+          },
+        ],
+      }),
```
Let's add `createdAt` and `updatedAt` columns to this table.

Same as how users is setup:
https://github.com/ericdowell/rerun/blob/ea18b97e0241c14f19d52afbfec95ef26b344dfb/src/migration/1593317181680-CreateUsersTable.ts#L35-L46

larzhole replied:
Sounds good.
Could we also maybe create our own BaseEntity that has those and just extend from that

ericdowell replied:
Yup. Sounds Gouda. You can import `typeorm` `BaseEntity` either default or aliasing it's name via `as`:
```ts
// Default usage
import typeorm from 'typeorm'

class BaseEntity extends typeorm.BaseEntity {}

// Or via as aliasing

import { BaseEntity as TypeORMBaseEntity from 'typeorm'

class BaseEntity extends TypeORMBaseEntity {}
```

larzhole replied:
Complete. I'm more of a fan of the aliasing option for some reason so that's what I went with.

ericdowell commented on src/migration/1594835753320-CreateUserRolesTable.ts:
```diff
@@ -0,0 +1,55 @@
+            isPrimary: true,
+            isGenerated: false,
+            isNullable: false,
+          },
```
Let's add `createdAt` and `updatedAt` columns to this table.

Same as how users is setup:
https://github.com/ericdowell/rerun/blob/ea18b97e0241c14f19d52afbfec95ef26b344dfb/src/migration/1593317181680-CreateUsersTable.ts#L35-L46

larzhole replied:
Complete

ericdowell commented on src/app.ts:
```diff
@@ -134,6 +139,51 @@ const registerMiddleware = (app: Express): void => {
+      return next()
+    }
+
+    const currentUserRoles = user?.userRoles?.map((userRole) => userRole.role.name)
+
+    if (!currentUserRoles) {
+      return res.status(401).json({ message: 'This action is unauthorized.' })
+    }
```
We can update this to allow `guest` acl:
```suggestion
     const currentUserRoles = user?.userRoles?.map((userRole) => userRole.role.name) ?? []
     const currentRoles [...currentUserRoles, 'guest'].
```
No need to check if `currentUserRoles` is empty since all activity will be given a `guest` role.

This will keep the normal logic for getting the user roles and then use the spread operator to add them into a new array that will also contain `guest`.

ericdowell commented on src/app.ts:
```diff
@@ -134,6 +139,51 @@ const registerMiddleware = (app: Express): void => {
+  // Access Control List
+  app.use(async (req: Request, res: Response, next: NextFunction) => {
+    const allRoutes: string[] = Object.values(Routes)
+    const guestRoutes: string[] = [
```
Let's move this `guestRoutes` to the `src/auth/acl.ts` as the `guest` entry.

ericdowell replied:
We removed `guest` authorizations because we will now apply the acl middleware at a route by route.

ericdowell commented on src/auth/acl.ts:
```diff
@@ -0,0 +1,20 @@
+      delete: [],
+    },
+  },
+  {
```
We can add a `guest` entry that all non-logged in users will get applied to.

Keeping the keys in alphabetical order.

ericdowell replied:
Updated, `guest` will not have any authorization power.

ericdowell commented on src/app.ts:
```diff
@@ -134,6 +139,51 @@ const registerMiddleware = (app: Express): void => {
+
+  // Access Control List
+  app.use(async (req: Request, res: Response, next: NextFunction) => {
+    const allRoutes: string[] = Object.values(Routes)
```
This does take into account `method`. We can work on this later though, not a big deal. Likely we'll make route more dynamic so we'll know which routes have what method registered.

A better name for this variable is `registeredRoutes`.

ericdowell commented on src/migration/1594835753320-CreateUserRolesTable.ts:
```diff
@@ -0,0 +1,55 @@
+            type: 'bigint',
+            unsigned: true,
+            isPrimary: true,
+            isGenerated: false,
```
Should be:
```suggestion
            isGenerated: true,
            generationStrategy: 'increment',
```

larzhole replied:
You must have meant this for the Roles table. I will make the change there but UserRoles should not be generated

ericdowell commented on src/__tests__/__data__/user.ts:
```diff
@@ -12,6 +14,8 @@ export const generateUser = async (attr?: {
   user.email = attr?.email ?? 'foo.bar34@example.com'
   user.name = attr?.name ?? 'Foobar Doe'
+  const role: Role = generateRole()
+  user.userRoles = generateUserRoles({ user, roles: new Array(role) })
```
```ts
   user.userRoles = generateUserRoles({ user, roles: [role] }) 
```

ericdowell commented on src/app.ts:
```diff
@@ -79,18 +84,16 @@ const registerMiddleware = (app: Express): void => {
   // Authenticate Request using session.userId
   app.use(async (req: Request, res: Response, next: NextFunction) => {
```
We can move all of these middleware functions into separate files within the `src/middleware` folder.

ericdowell commented on src/entity/User.ts:
```diff
@@ -31,15 +44,9 @@ export class User extends BaseEntity {
-  @OneToOne((type) => PasswordReset, (passwordReset: PasswordReset) => passwordReset.user)
-  passwordReset: PasswordReset
+  getRoles(): string[] {
+    return this.userRoles?.map((userRole) => userRole.role.name) ?? ['user']
```
We'll default `user` if `userRoles` are not defined. Just a note in case some discussion is needed or this is incorrectly being set.

The idea is that anything calling this is a valid `User` model and thus is signed in.

ericdowell commented on src/app.ts:
```diff
@@ -79,18 +84,16 @@ const registerMiddleware = (app: Express): void => {
-      req.session.userId = user.id
       req.user = user
+      req.session.userId = user.id
+      req.session.roles = user.getRoles()
```
This might not be working. In that if the `user.userRoles` is not loaded, the `getRoles` method will default to `user`, which may be incorrect.

This would only happen during login where we are not using `user = await User.findOneOrFail(req.session.userId, { relations: ['userRoles'] })` within calling controller.

If there is a TypeORM method that allow for us to `loadMissing` relations, that would be ideal here. Otherwise we should make sure this is always eager loaded and thus never worry.

ericdowell replied:
Removed because storing this in the session is just one more place to maintain state and thus we could have stale data. That will likely create more bugs than benefit us from querying the database too much.

larzhole commented on src/controllers/user.ts:
```diff
@@ -50,6 +50,28 @@ export const putUpdateUser = async (req: Request, res: Response): Promise<Respon
+  }
+
+  try {
+    const userRole: UserRole = req.body
```
Good call, fixed

larzhole commented on src/controllers/user.ts:
```diff
@@ -50,6 +50,28 @@ export const putUpdateUser = async (req: Request, res: Response): Promise<Respon
+
+export const patchUserRole = async (req: Request, res: Response): Promise<Response> => {
+  await check('userId', 'User ID is required.').exists().run(req)
+  await check('userId', 'User ID must be numeric.').not().isString().run(req)
```
Fixed

larzhole commented:
> Let's add in the `todo` existing queries I added to RoleController and add tests for these blocks of code:
> https://github.com/ericdowell/rerun/blob/2d6103ad4a8034141eef0be0803791a0a7d5c487/src/controllers/role.ts#L48-L62
> 
> 
> https://github.com/ericdowell/rerun/blob/2d6103ad4a8034141eef0be0803791a0a7d5c487/src/controllers/role.ts#L99-L113

Fixed. Although I'm not a fan of empty catches, seems like an anti-pattern.
When I start on the Error Response work I'll look into it some more
